Github pages (#31)
Supercolony-net/openbrush-contracts: doc/src/EVM-WASM-Smart-Contracts.md

##### Gas model
The incentive model of interacting with EVM is gas. It acts like a fee that is calculated for each instruction you are executing. Every instruction has a certain pre-calculated fee amount, and your total gas is the sum of all the instruction you will execute.  

Please check the entire table of [fees](https://blockgeeks.com/wp-content/uploads/2018/03/image2-2.png)
A gas fee is only due when interacting with EVM. There is no charge for the space on storage you are using, and it is resulting by having a ton of deadcode (or non-used contracts) staying on ethereum chain forever.

##### Evm process

![image info](pictures/eth2.png)

### Substrate Contract-Pallet

##### Intro
`pallet-contract` is a module in Substrate (the framework to build blockchain on `Polkadot`). Its purpose is a sandbox environment providing WASM interpreter to execute smart contracts in WASM.

![image info](pictures/WASM1.png)

##### WASM interpreter

At the moment contract-pallet use [wasmi](https://github.com/paritytech/wasmi) as interpreter. Wasmi is a pure interpreter-type WASM virtual machine as execution of smart contract needs a high degree of correctness.

##### Storage rent & Gas

> In order to incentives the deletion of unused code on chain, pallet-contract implemented a storage-rent principle.  

Reviewer: ```suggestion
In order to incentive the deletion of unused code on chain, pallet-contract implemented a storage-rent principle.  
```

> Data that persist on chain (the contract code + all of its storage) will be charged. So every smart contract will have a rent to pay (small amount at every block). When the smart contract will not have enough fund to pay the rent the smart contract will become a Tombstone (storage will be deleted).  

Reviewer: ```suggestion
Every smart contract will have a rent to pay for its code and on-chain storage (small amount at every block). When the smart contract fails to pay the rent, the contract will become a Tombstone (its storage will be deleted).  
```

> The gas system of ethereum (price depending on the complexity of the computation) is still present, but it is charged after the call is executed as it is basically a fee on the time of execution (the more time it takes for the node to execute your call the more you will pay). The `pallet-contract` will define the [amount of gas](https://substrate.dev/docs/en/knowledgebase/smart-contracts/contracts-pallet)

Reviewer: ```suggestion
The gas system of ethereum (price depending on the complexity of the computation) is still present in Substrate, but the gas is charged after the call is executed, as it is basically a fee on the time of execution (the more time it takes for the node to execute a call, the more caller will pay). The `pallet-contract` defines the [amount of gas](https://substrate.dev/docs/en/knowledgebase/smart-contracts/contracts-pallet)
```

##### Contract code & instance are decoupled

> Even though pallet-contract use an account model for their contracts alike ethereum does as well, there is still a big difference:  

Reviewer: ```suggestion
Even though `pallet-contract` uses an account model for its contracts (like ethereum does), there is still one big difference:  
```

> When you deploy a WASM smart contract on chain it will only create a hash of the WASM code, that's mean this contract will not have an address, nor an associated storage (so no interaction possible). As opposed to ethereum where every contract code deployed on chain will have a unique address (instance), and an associated storage.  

Reviewer: ```suggestion
When you deploy a WASM smart contract, it will only create a hash of the WASM code, meaning that this contract will have no address, nor will it have some associated storage (so it is impossible to interact with this contract). In Ethereum, though, every contract code deployed on chain has a unique address (instance) and its associated storage.  
```

> When the WASM hash is on chain you can create as mush as instance of this contract code. Each instance will have a unique address to interact with as well as its own associated storage.  

Reviewer: ```suggestion
After the WASM hash is deployed, one can create as many instances of this contract as they want. Each instance will have a unique address to interact with, as well as its own associated storage.  
```

> It is practical for:
> - Different contract instance with different **constructors** can be instantiated to reduce the space needed on chain to store WASM code
> - Storage and balance are decoupled from contract code logic, it can enable to **patch or upgrade** the underlying contract code

Reviewer: ```suggestion
Why is it useful:
- Different contract instances with different **constructor parameters** can be instantiated using the same uploaded code, which reduces the space needed on chain to store WASM code
- Storage and balance are decoupled from contract code logic, which provides ability **patch or upgrade** the underlying contract code
```

### EVM vs contract-pallet

- Common point: they both are a sandbox to execute smart contracts

> - The engine to execute contract is different. In ethereum, as it was the first blockchain to implement n sandbox environment, it is rather limited and slow compare to a wasm interpreter

Reviewer: ```suggestion
- The engine which executes contracts is different. In Ethereum, as it was the first blockchain to implement a sandbox environment, it is rather limited and slow, compared to a wasm interpreter
```

> - Storage-rent has been introduced in contract-pallet to incentives the deletion of unused code

Reviewer: ```suggestion
- Storage rent has been introduced in contract-pallet to incentivize the deletion of unused code
```

> - Contract pallet integrated a [two-step-deployment](https://substrate.dev/docs/en/knowledgebase/smart-contracts/contracts-pallet#two-step-deployment) to decouple contract code and contract instances

Reviewer: ```suggestion
- Contract pallet integrates a [two-step-deployment](https://substrate.dev/docs/en/knowledgebase/smart-contracts/contracts-pallet#two-step-deployment) to decouple contract code and contract instances
```

### Comparing WASM Smart contracts to EVM smart contracts

> - WASM is broadly adopted as EVM usage is only for EVM -> There is way more tools available for WASM development
> - EVM bytecode can only be compiled from Solidity or Vyper while WASM can be compiled from a lot of Popular languages (Rust, C/C++, C#, Java, Typescript, Haxe, Kotlin and even from Solidity)

Reviewer: ```suggestion
- WASM is broadly adopted, while EVM bytecode is only used in EVMs -> There is way more tools available for WASM development
- EVM bytecode can only be compiled from Solidity or Vyper, while WASM can be compiled from a lot of Popular languages (Rust, C/C++, C#, Java, Typescript, Haxe, Kotlin and even from Solidity)
```
